File: android/app/src/main/java/com/aerion/tablefeed/ui/discover/MealClassifier.kt
package com.aerion.tablefeed.ui.discover

enum class MealType(val label: String) {
    BREAKFAST("Breakfast"), LUNCH("Lunch"), DINNER("Dinner"), SNACK("Snack"), DESSERT("Dessert")
}

/**
 * Heuristic meal-type classification from recipe tags (no schema change — works on existing data).
 * A recipe can match multiple meals; an unclassifiable recipe is eligible for any slot.
 */
object MealClassifier {

    private val keywords: Map<MealType, List<String>> = mapOf(
        MealType.BREAKFAST to listOf("breakfast", "brunch", "pancake", "waffle", "omelet", "omelette",
            "egg", "oats", "oatmeal", "cereal", "toast", "smoothie", "granola"),
        MealType.DESSERT to listOf("dessert", "cookie", "cake", "brownie", "sweet", "chocolate",
            "ice cream", "pudding", "pie", "pastry", "cupcake", "candy", "cheesecake"),
        MealType.SNACK to listOf("snack", "appetizer", "starter", "dip", "finger food", "nibble"),
        // Savory mains serve both lunch and dinner.
        MealType.DINNER to listOf("dinner", "lunch", "main", "curry", "pizza", "pasta", "rice",
            "biryani", "soup", "stew", "sandwich", "burger", "salad", "roast", "grill",
            "noodle", "stir fry", "taco", "bowl"),
    )

    /** All meal keywords flattened — used to exclude meal words when inferring cuisine affinity. */
    val allMealKeywords: Set<String> = keywords.values.flatten().toSet()

    fun mealsFor(tags: List<String>): Set<MealType> {
        val hay = tags.joinToString(" ").lowercase()
        val result = mutableSetOf<MealType>()
        keywords.forEach { (meal, kws) -> if (kws.any { hay.contains(it) }) result.add(meal) }
        if (MealType.DINNER in result) result.add(MealType.LUNCH) // savory mains → both
        return result
    }

    /** Map an hour-of-day (0–23) to the meal slot to lead the feed with. */
    fun currentMeal(hour: Int): MealType = when (hour) {
        in 5..10 -> MealType.BREAKFAST
        in 11..14 -> MealType.LUNCH
        in 15..16 -> MealType.SNACK
        in 17..21 -> MealType.DINNER
        else -> MealType.DESSERT
    }
}

File: ios/ChefsJournal/ChefsJournal/UI/Discover/MealClassifier.swift
import Foundation

enum MealType: String, CaseIterable {
    case breakfast = "Breakfast"
    case lunch = "Lunch"
    case dinner = "Dinner"
    case snack = "Snack"
    case dessert = "Dessert"
    var label: String { rawValue }
}

/// Heuristic meal-type classification from recipe tags (no schema change — works on existing data).
/// A recipe can match multiple meals; an unclassifiable recipe is eligible for any slot.
enum MealClassifier {
    private static let keywords: [MealType: [String]] = [
        .breakfast: ["breakfast", "brunch", "pancake", "waffle", "omelet", "omelette",
                     "egg", "oats", "oatmeal", "cereal", "toast", "smoothie", "granola"],
        .dessert: ["dessert", "cookie", "cake", "brownie", "sweet", "chocolate",
                   "ice cream", "pudding", "pie", "pastry", "cupcake", "candy", "cheesecake"],
        .snack: ["snack", "appetizer", "starter", "dip", "finger food", "nibble"],
        // Savory mains serve both lunch and dinner.
        .dinner: ["dinner", "lunch", "main", "curry", "pizza", "pasta", "rice",
                  "biryani", "soup", "stew", "sandwich", "burger", "salad", "roast", "grill",
                  "noodle", "stir fry", "taco", "bowl"]
    ]

    /// All meal keywords — used to exclude meal words when inferring cuisine affinity.
    static let allMealKeywords: Set<String> = Set(keywords.values.flatMap { $0 })

    static func mealsFor(_ tags: [String]) -> Set<MealType> {
        let hay = tags.joined(separator: " ").lowercased()
        var result = Set<MealType>()
        for (meal, kws) in keywords where kws.contains(where: { hay.contains($0) }) {
            result.insert(meal)
        }
        if result.contains(.dinner) { result.insert(.lunch) } // savory mains → both
        return result
    }

    /// Map an hour-of-day (0–23) to the meal slot to lead the feed with.
    static func currentMeal(hour: Int) -> MealType {
        switch hour {
        case 5...10: return .breakfast
        case 11...14: return .lunch
        case 15...16: return .snack
        case 17...21: return .dinner
        default: return .dessert
        }
    }
}
